From c4322e8e24305e9f68ad7a675fa132d222773817 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 5 Sep 2019 09:50:23 +0200
Subject: [PATCH] PONRTSYS-5112: Allow relocation for kernel on IAP

This makes the IAP build with CONFIG_SYS_SUPPORTS_RELOCATABLE
and still finding the appended dtb.
---
 arch/mips/lantiq/Kconfig       |  2 ++
 arch/mips/lantiq/grx500/prom.c | 14 ++++++++------
 2 files changed, 10 insertions(+), 6 deletions(-)

diff --git a/arch/mips/lantiq/Kconfig b/arch/mips/lantiq/Kconfig
index 41fc5c3a3ad8..4ec7f8cd36d8 100755
--- a/arch/mips/lantiq/Kconfig
+++ b/arch/mips/lantiq/Kconfig
@@ -57,6 +57,8 @@ config SOC_GRX500
 	select CPU_SUPPORTS_CPUFREQ
 	select MIPS_EXTERNAL_TIMER
 	select PCI_DRIVERS_GENERIC
+	select SYS_SUPPORTS_RELOCATABLE
+
 config SOC_GRX500_BOOTCORE
 	bool "GRX500_BOOTCORE"
 	select SOC_TYPE_GRX500_TEP
diff --git a/arch/mips/lantiq/grx500/prom.c b/arch/mips/lantiq/grx500/prom.c
index 2e710cf7a01f..0cd94a99b2fb 100755
--- a/arch/mips/lantiq/grx500/prom.c
+++ b/arch/mips/lantiq/grx500/prom.c
@@ -30,6 +30,7 @@
 /* access to the ebu needs to be locked between different drivers */
 DEFINE_SPINLOCK(ebu_lock);
 EXPORT_SYMBOL_GPL(ebu_lock);
+static __initdata const void *dtb;
 
 /*
  * this struct is filled by the soc specific detection code and holds
@@ -142,10 +143,8 @@ static void free_init_pages_eva_xrx500(void *begin, void *end)
 			__pa_symbol((unsigned long *)end));
 }
 
-static void plat_early_init_devtree(void)
+void __init *plat_get_fdt(void)
 {
-	void *dtb;
-
 	/*
 	 * Load the builtin devicetree. This causes the chosen node to be
 	 * parsed resulting in our memory appearing
@@ -157,8 +156,7 @@ static void plat_early_init_devtree(void)
 	else
 		panic("no dtb found");
 
-	if (dtb)
-		__dt_setup_arch(dtb);
+	return (void *)dtb;
 }
 
 void __init plat_mem_setup(void)
@@ -184,7 +182,9 @@ void __init plat_mem_setup(void)
 
 	set_io_port_base((unsigned long)KSEG1);
 
-	plat_early_init_devtree();
+	if (dtb)
+		__dt_setup_arch((void *)dtb);
+
 	plat_setup_iocoherency();
 	free_init_pages_eva = eva ? free_init_pages_eva_xrx500 : NULL;
 }
@@ -250,6 +250,8 @@ void __init prom_init(void)
 	pr_info("SoC: %s\n", get_system_type());
 
 	prom_init_cmdline();
+	plat_get_fdt();
+	BUG_ON(!dtb);
 
 	mips_cpc_probe();
 
